learning_observer/learning_observer/canvas.py

```diff
@@ -0,0 +1,148 @@
+            self.access_token = self.config['CANVAS_CONFIG']['ACCESS_TOKEN']
+            self.refresh_token = self.config['CANVAS_CONFIG']['REFRESH_TOKEN']
+            self.client_id = self.config['CANVAS_CONFIG']['CLIENT_ID']
+            self.client_secret = self.config['CANVAS_CONFIG']['CLIENT_SECRET']
+        except KeyError as e:
+            raise KeyError(f"Missing required configuration key: {e}")
+        self.default_version = 'v1'
+        self.defaultPerPage = 10000
+        self.base_url = f'https://{self.defaultServer}/api/{self.default_version}'
+
+    def update_access_tokens(self, access_token):
+        self.config['CANVAS_CONFIG']['ACCESS_TOKEN'] = access_token
+        self.access_token = access_token
+        script_dir = os.path.dirname(os.path.abspath(__file__))
+        config_path = os.path.join(script_dir, './config.ini')
+        with open(config_path, 'w') as configfile:
+            self.config.write(configfile)
+
+    async def api_call(self, method, endpoint, params=None, data=None, absolute_url=False, retry=True, **kwargs):
+        if absolute_url:
+            url = endpoint
+        else:
+            url = self.base_url + endpoint
+            #if params:
+                #url += '?' + '&'.join(f"{k}={v}" for k, v in params.items())
+        
+        url = url.format(**kwargs)
+
+        headers = {
+            'Authorization': f'Bearer {self.access_token}',
+            'Content-Type': 'application/json'
+        }
+
+        async with aiohttp.ClientSession() as client:
+            response_func = getattr(client, method.lower())
+            async with response_func(url, headers=headers, params=params, json=data) as response:
+                if response.status == 401 and retry:
+                    new_tokens = await self.refresh_tokens()
+                    if 'access_token' in new_tokens:
+                        self.update_access_tokens(new_tokens['access_token'])
+                        return await self.api_call(method, endpoint, params, data, absolute_url, retry=False, **kwargs)
+                
+                if response.status != 200:
+                    response.raise_for_status()
+                
+                return await response.json()
+
+    async def refresh_tokens(self):
+        url = f'https://{self.defaultServer}/login/oauth2/token'
+        params = {
+            "grant_type": "refresh_token",
+            "client_id": self.client_id,
+            "client_secret": self.client_secret,
+            "refresh_token": self.refresh_token
+        }
+        return await self.api_call('POST', url, params=params, absolute_url=True)
+
+        
+class CanvasLMS(BaseLMS):
+    def __init__(self):
+        super().__init__(lms_name="canvas", endpoints=CANVAS_ENDPOINTS, raw_ajax_function=self.raw_canvas_ajax)
+        self.canvas = Canvas()
+
+    async def raw_canvas_ajax(self, runtime, target_url, retry=False, **kwargs):
```

Comment:
This and the Google code look very similar. We want to avoid similar code. The Google code does a lot of extra steps for caching results if specific settings are present. Each LMS should probably do this.

Since Google was the first LMS implemented, many variables include google in their name. The purpose of the `use_google_ajax` is to save responses from the Google API to the local filesystem for debugging purposes. This would be a great functionality for the Canvas items.

Perhaps the LMS class should have a KVS ready.

Reply:
I've abstracted the raw_ajax code for all LMS